feat(symfony): Autoconfigure resource dirs using `#[ApiResource]` attribute

GromNaN opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | main
| Tickets       | -
| License       | MIT
| Doc PR        | todo

Using the Symfony bundle, classes with the `#[ApiResource]` attribute must be in `src/ApiResource`, `src/Entity` (for Doctrine ORM) or `src/Document` (for MongoDB ODM). It's possible to add custom paths in the configuration, but that's one more little thing to take care of.

By using the Symfony autoconfiguration feature, we can detect classes with the `#[ApiResource]`  attribute during the build of the container instead of scanning directories at runtime.

GromNaN commented on src/Symfony/Bundle/DependencyInjection/ApiPlatformExtension.php:
```diff
@@ -168,6 +169,10 @@ public function load(array $configs, ContainerBuilder $container): void
             ->addTag('api_platform.parameter_provider');
+        $container->registerAttributeForAutoconfiguration(ApiResource::class, static function(ChildDefinition $definition): void {
+            $definition->addTag('api_platform.resource');
+            $definition->addTag('container.excluded', ['source' => __FILE__]);
```
Since ApiResource classes are not services but value objects, the tag `container.excluded` is also added, so they cannot be used in autowiring.

nicolas-grekas replied:
"source" is rather the file where the attribute is defined

nicolas-grekas replied:
(we could add it automatically BTW, PR welcome :)

GromNaN replied:
Instead of indicating the source file (which is easy to deduce from the class), we can explain that it is the use of this attribute that causes the class to be excluded.

```suggestion
            $definition->addTag('container.excluded', ['source' => 'with #[ApiResource] attribute']);
```

For context, this is used in [`AutowirePass`](https://github.com/symfony/symfony/blob/6f4f04b98f6134996a57d21d1851d558a34dc45c/src/Symfony/Component/DependencyInjection/Compiler/AutowirePass.php#L601).

GromNaN commented on src/Symfony/Bundle/DependencyInjection/Compiler/AttributeResourcePass.php:
```diff
@@ -0,0 +1,48 @@
+            if ($definition->hasTag('api_platform.resource')) {
+                $r = new \ReflectionClass($definition->getClass());
+                if ($r->getFileName()) {
+                    $resourceClassDirectories[] = \dirname($r->getFileName());
```
I found `api_platform.class_name_resources`, the correct parameter to set class names.

dunglas reviewed (approved):
Very smart! I like this.

soyuka commented:
Nice! It reminds me a discussion we had with @nicolas-grekas ^^. 
I'm wondering a few things, do we still need:

https://github.com/api-platform/core/blob/7b3bf838b31f51d47c03d2318fd9cb5c4dd419ba/src/Symfony/Bundle/DependencyInjection/ApiPlatformExtension.php#L342-L348

The way `api_platform.class_name_resources` works right now it's called by `api_platform.metadata.resource.name_collection_factory.class_name`, until this patch classes were discovered using `api_platform.resource_class_directories` inside `api_platform.metadata.resource.name_collection_factory.attributes`. Shouldn't we clean that part of the code to remove php files?

https://github.com/api-platform/core/blob/7b3bf838b31f51d47c03d2318fd9cb5c4dd419ba/src/Symfony/Bundle/DependencyInjection/ApiPlatformExtension.php#L385-L433 

This is quite heavy work for no real gain, especially that the resources discovered with this patch will already be discovered there right?

Last thing, inside the route loader we declare directory resources (so that the container is rebuilt when these file changes) at:

https://github.com/api-platform/core/blob/7b3bf838b31f51d47c03d2318fd9cb5c4dd419ba/src/Symfony/Routing/ApiLoader.php#L52-L54

If we remove that  `api_platform.resource_class_directories` for php files, we can also remove these lines but we may need to add something to handle these no?

GromNaN commented:
Removing the directory scanning would be a breaking change as the `Entity` directory is excluded by default from the services config (see [FrameworkBundle recipe](https://github.com/symfony/recipes/blob/2d8f656159ca9132c55b6e848171a7dabc1d8554/symfony/framework-bundle/7.2/config/services.yaml#L20)).

Of-course, this is the target.

soyuka commented:
Mhh okay but this means that we only need this for the Entity directory, we can remove all the code handling our paths (for example ApiResource). What about the DirectoryResource to rebuild the container on file change? Are services tagged as `container.excluded` watched by Symfony?

dunglas commented:
If this doesn’t work for the Entity directory (that we won’t deprecate IMHO), then is this patch really useful? 
We’ll have 2 systems to maintain basically forever instead of one 😅

GromNaN commented:
Let's reintroduce the `src/Entity` dir in the DI config: https://github.com/doctrine/DoctrineBundle/pull/1868

I'll try to move the class discovery to the compiler pass.

GromNaN commented:
So I think we need to move the extraction of the classes names into the compiler pass and inject them into a `ClassNameResourceNameCollectionFactory`. The `CachedResourceNameCollectionFactory` can then be removed.

GromNaN commented:
Here's how it should be designed to take full advantage of the container builder.

### Goals
- Enable auto-discovery of `#[ApiResource]` classes anywhere
- Build the list of "resource" classes during container compilation

**Advantage:** no need for additional cache unless explicitly required by a custom implementation `CachedResourceNameCollectionFactory`.

### Backward-compatibility constraints:
- `src/ApiResource`, `src/Entity`, `src/Document` directories must be scanned even if they are not in the DIC autoconfigure paths. This can be deprecated.
- Keep support for custom `ResourceNameCollectionFactoryInterface` services
- Keep support for XML/Yaml files: Detection should be similar to any other config files (doctrine.yaml, validation.yaml with [`registerMappingFilesFromConfig`](https://github.com/symfony/symfony/blob/86c8a8ab43e3118dfe213ca1db635e9a0223e7b5/src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php#L1802)

### Attributes

**Detection:** Use Symfony `ContainerBuilder::registerAttributeForAutoconfiguration()` to list all the classes that use the `#[ApiResource]` attribute and inject them into a `ClassNameResourceNameCollectionFactory`.

**Cache invalidation:** The Symfony cache is automatically invalidated and regenerated when one of this classes is modified, removed or added in the autoconfigured directories.

### XML/Yaml files

**Detection:**: List files during container compilation.

**Cache invalidation:** Adding this files and directories as "resources"; even if they doesn't exist, the container cache is automatically invalidated and regenerated when on of the files is modified, removed or added in the configured directories.

Nayte91 commented:
Hello, I may miss a link in this discussion, but I didn't see here anything related to https://github.com/symfony/symfony/pull/59987 as for SF7.3 #[Entity] tagged classes will be automatically excluded from services, and therefore default config.yaml services definition folders will be cleaned.

Sorry if already mentionned.

soyuka commented:
For now we can merge this as-is, we'll deprecate the `resource_class_directories`, `mapping.paths` and add a flag to skip resource class directories.

michaljusiega commented on src/Symfony/Bundle/DependencyInjection/Configuration.php:
```diff
@@ -141,6 +141,7 @@ public function getConfigTreeBuilder(): TreeBuilder
                 ->end()
                 ->arrayNode('resource_class_directories')
                     ->prototype('scalar')->end()
+                    ->setDeprecated('api-platform/symfony', '4.1', 'The "resource_class_directories" configuration is deprecated, classes using #[ApiResource] attribute are autoconfigured by the dependency injection container.')
```
@soyuka @GromNaN 

I think, this feature break some functionality.

If this option was deprecated, and planned to be removed in future, this will break any additional features like filters (sorting/filtering) on the same level as `#[ApiResource]` in resource - this means they are not loaded with all the operations they are should be.

Simple reproducer:
`api-platform/core:4.2.x-dev`

```
<?php

use ApiPlatform\Doctrine\Common\Filter\SearchFilterInterface;
use ApiPlatform\Doctrine\Orm\Filter\SearchFilter;
use ApiPlatform\Metadata\ApiFilter;
use ApiPlatform\Metadata\ApiResource;
use ApiPlatform\Metadata\GetCollection;
use ApiPlatform\OpenApi\Model\Operation;
use App\Expense\Application\Controller\GetCollectionCostController;
use App\Shared\Domain\Entity\AbstractEntity;

#[ApiResource(
    operations: [
        new GetCollection(
            uriTemplate: 'expense/cost',
            inputFormats: ['jsonld' => ['application/ld+json']],
            outputFormats: ['jsonld' => ['application/ld+json']],
            controller: GetCollectionCostController::class,
            openapi: new Operation(tags: ['Expense'], summary: 'Returns the collection of Cost resources.', description: 'Returns the collection of Cost resources.'),
            normalizationContext: ['groups' => ['api:read']],
            denormalizationContext: ['groups' => ['api:write']],
            output: Cost::class,
            priority: 2,
        ),
    ],
)]
#[ApiFilter(filterClass: SearchFilter::class, properties: ['comment' => SearchFilterInterface::STRATEGY_IPARTIAL])]
class Cost extends AbstractEntity
{
}
```

With setting `resource_class_directories`:
![obraz](https://github.com/user-attachments/assets/1d205815-482a-49cd-b0a6-b3cd39c24bed)

and deprecated message:
`Since api-platform/symfony 4.1: The "resource_class_directories" configuration is deprecated, classes using #[ApiResource] attribute are autoconfigured by the dependency injection container.`

Without setting `resource_class_directories`:
![obraz](https://github.com/user-attachments/assets/5fc2f8e3-13de-4984-992f-cba668877dee)

I don't know how to fix it. Unless there is some other way to fix it?

michaljusiega replied:
Settings the `#[ApiFilter]` attribute directly on property (like in https://api-platform.com/docs/core/filters/#apifilter-attribute) also not works.

michaljusiega replied:
Do you have any information about that @soyuka ?

GromNaN replied:
Should other attributes such as `#[ApiResource]` be added to the AutoConfiguration?

michaljusiega replied:
@GromNaN I think, yes.

soyuka replied:
yes ApiFilter should also be auto configured I'm not sure when I'll have the time to handle this

soyuka replied:
I think we're going to deprecate ApiFilter at some point therefore not sure its worth it to autoconfigure it.

soyuka replied:
https://api-platform.com/docs/core/filters/

michaljusiega replied:
I tried to achive the same result as is in https://api-platform.com/docs/core/filters/#filtering-multiple-properties-with-property and I failed...

Cannot instance new `SearchFilter` like wrote in docs (because of required arguments like ManagerRegistry)...

So for me +1 to keep `ApiFilter` class to be not deprecated.

michaljusiega replied:
Thank you very much for response. I already know what way further development goes.
